beetsplug/kodinfo.py

```diff
@@ -0,0 +1,448 @@
+        else:
+            folder = os.path.join(root, str(config['paths']['default']))
+            config_items = Path(folder).parts
+            folder_length = len(config_items)
+            indices = [i for i,
+                       s in enumerate(config_items) if 'albumartist' in s]
+            y = int(folder_length - indices[0])
+            artistpath_items = list(Path(albumpath).parts[-y:-1])
+            artist_path = os.path.join(root, *artistpath_items)
+            kodi_lib_path = kodi_path() + artist_path[int(len(root) + 1):]
+        out_data = artist_path, kodi_lib_path
+    return out_data
+
+
+def thumbs(tag, albumid):
+    """Name paths where art files reside."""
+    if "artist" in tag:
+        thumbs = []
+        for a in artist_path(albumid):
+            thumb = os.path.join(a, 'artist.tbn')
+            thumbs.append(thumb)
+        return thumbs
+
+    if "album" in tag:
+        for album in lib.albums(albumid):
+            if album.artpath:
+                art_file = os.path.basename(album.artpath.decode('utf8'))
+        thumbs = []
+        for a in album_path(albumid):
+            thumb = os.path.join(a, art_file)
+            thumbs.append(thumb)
+        return thumbs
+
+
+def album_nfo_text(albumid, mb_albumid, mb_artistid):
+    """Create MBID URL only text file."""
+    album_nfo_file = os.path.join(album_path(albumid)[0], 'album.nfo')
+    with open(album_nfo_file, 'w') as f:
+        f.write(LINK_ALBUM.format(mb_albumid))
+    if os.path.basename(artist_path(albumid)[0]) in ['Various Artists',
+                                                     'Soundtracks',
+                                                     'Compilations']:
+        pass  # No artist.nfo file for compilation albums
+    else:
+        artist_nfo_file = os.path.join(artist_path(albumid)[0],
+                                       'artist.nfo')
+        with open(artist_nfo_file, 'w') as f:
+                f.write(LINK_ARTIST.format(mb_artistid))
+
+
+def album_nfo_xml(albumid):
+    """Create XML file with album information."""
+    from lxml import etree as et
+    for album in lib.albums(albumid):
+        albumnfo = os.path.join(album.path.decode('utf8'), 'album.nfo')
+        albumid = 'mb_albumid:' + album.mb_albumid
+        root = et.Element('album')
+        for i in range(len(album_tags)):
+            album_tags[i] = et.SubElement(root, '{}'.format(album_tags[i]))
+            album_tags[i].text = album_info(albumid)[i]
+
+        for i in range(len(album_path(albumid))):
+            path = et.SubElement(root, 'path')
+            path.text = album_path(albumid)[i]
+
+        if album_info(albumid)[14] == '':
+            for i in range(len(thumbs('album', albumid))):
+                thumb = et.SubElement(root, 'thumb')
+                thumb.text = thumbs('album', albumid)[i]
+        else:
+            thumb = et.SubElement(root, 'thumb')
+            thumb.text = album_info(albumid)[14]
```

Comment:
This looks like it calls `album_info` three times, which will fetch the same result from the metadata server three times. Maybe it would be wise to store the result and reuse it? (Or, this is part of why I'm suggesting that this part of the plugin be moved somewhere else—the data from the server can just be stored in the database.)